### DISCUSSION TITLE
Executing R scripts from Dagster

### DISCUSSION THREAD

**Discussion by philiporlando** [dagster-io/dagster#18972] [Q&A] [answered]
I'm currently exploring Dagster for data orchestration and find myself in need of some community wisdom. My primary goal is to execute R scripts within the Dagster environment, a task I previously accomplished using Airflow's BashOperator and DockerOperator.

In Airflow, these operators provided a straightforward way to run R scripts either directly through shell commands or within Docker containers. I'm looking for Dagster's equivalents to these operators. Specifically, I'm interested in:

1. How to execute R scripts in Dagster, comparable to using Airflow's BashOperator. I'd like to understand the best practices for running shell commands or scripts, especially for R scripts.

```python
from airflow import DAG
from airflow.operators.bash_operator import BashOperator
from datetime import datetime

default_args = {
    'owner': 'airflow',
    'start_date': datetime(2023, 1, 1),
    # other args...
}

dag = DAG('r_script_bash_operator', default_args=default_args, schedule_interval='@daily')

run_r_script = BashOperator(
    task_id='run_r_script',
    bash_command='Rscript /path/to/your_script.R ',
    dag=dag,
)

run_r_script
```

2. The Dagster equivalent of Airflow's DockerOperator. My use case often requires running scripts in isolated environments, and I'm curious about executing R scripts within Docker containers managed by Dagster.

```python
from airflow import DAG
from airflow.operators.docker_operator import DockerOperator
from datetime import datetime

default_args = {
    'owner': 'airflow',
    'start_date': datetime(2023, 1, 1),
    # other args...
}

dag = DAG('r_script_docker_operator', default_args=default_args, schedule_interval='@daily')

run_r_script_in_docker = DockerOperator(
    task_id='run_r_script_in_docker',
    image='r-base',  # Make sure this image has R installed
    command='Rscript /path/to/your_script.R',
    docker_url='unix://var/run/docker.sock',  # Update as needed
    network_mode='bridge',
    dag=dag,
)

run_r_script_in_docker
```

I would greatly appreciate any examples, tips, or pointers to relevant documentation. Insights on capturing logging, metadata, and results while executing these scripts would also be incredibly helpful.

Thank you in advance for your guidance and support.

----

This is related to #2624.

**Comment by PedramNavid** [comment, marked as answer]
Hi @philiporlando and fellow R-user. 

You have a few options, but I think the simplest would be to use the [dagster-docker op](https://docs.dagster.io/_apidocs/libraries/dagster-docker#dagster_docker.docker_container_op).


From the docs: 

```python
from dagster_docker import docker_container_op

from dagster import job

first_op = docker_container_op.configured(
    {
        "image": "busybox",
        "command": ["echo HELLO"],
    },
    name="first_op",
)
second_op = docker_container_op.configured(
    {
        "image": "busybox",
        "command": ["echo GOODBYE"],
    },
    name="second_op",
)

@job
def full_job():
    second_op(first_op())
```

For more complex integrations, you might find Dagster Pipes interesting as well: https://docs.dagster.io/guides/dagster-pipes

Let me know if this is helpful or if you have other questions.

**Reply by philiporlando** [reply]
Thank you for pointing me in the right direction and for providing this example!

**Reply by philiporlando** [reply]
@PedramNavid, is it possible to point the `docker_container_op` to a local Docker image, or does this need to be hosted somewhere like DockerHub?

**Reply by philiporlando** [reply]
## Dagster + Docker + R

I've been troubleshooting this further and wanted to document my progress here. 

I'm realizing that a custom Docker image is needed in order to pass any custom business logic written in R to the `Rscript` binary within the `rocker/r-base` image. 

Here is a simple Docker image that I've published for testing purposes:

```docker
# Dockerfile
# Use the rocker/r-base image as the base image
FROM rocker/r-base

# Copy the test.R script from your local project directory to the image
COPY test.R /test.R

# Set the working directory to where test.R is located
WORKDIR /

# The default command to run when the container starts
# Dagster will run this command directly 
# CMD ["Rscript", "test.R"] 
```

Then, I created an R script called `test.R` at the root of my dagster project that looked like this:

```R
# test.R
print("Hello, World!")
``` 

Next, I defined my job using the `docker_container_op` like so:

```python
# dagster_and_r/test_docker.py
from dagster_docker import docker_container_op

from dagster import job

first_op = docker_container_op.configured(
    {
        "image": "philiporlando/test-dagster-and-r:latest",
        "command": ["Rscript", "test.R"],
    },
    name="first_op",
)


second_op = docker_container_op.configured(
    {
        "image": "philiporlando/test-dagster-and-r:latest",
        "command": ["Rscript", "test.R"],
    },
    name="second_op",
)


@job
def full_job():
    second_op(first_op())
```

Lastly, I needed to add my job to the root `__init__.py` within my module (`dagster_and_r`):

```python
# dagster_and_r/__init__.py
from dagster import Definitions, load_assets_from_modules

from . import test_docker

defs = Definitions(
    assets=all_assets,
    jobs=[test_docker.full_job],
)
```

From here, I was able to launch the dagster UI via `dagster dev` and  from manually start a new job run from the Launchpad!

![image](https://github.com/dagster-io/dagster/assets/24445174/e3e1f42c-0be1-4136-83db-02146db2d464)

I'll admit that this isn't the most ergonomic approach, and I'd much rather be writing all my pipelines using the software-defined asset abstraction, but this might be enough to help me port some legacy R projects over to Dagster to buy me time before I do a complete rewrite. 

One thing worth mentioning here is that I haven't figured out how to validate that the R script ran successfully. I'm not seeing "Hello, World!" getting printed anywhere on the dagster side. Not sure how to account for this without passing context between Dagster and R...

**Reply by philiporlando** [reply]
## Dagster Pipes + R

I was also able to get up and running with Dagster Pipes and R. 

First, I created a new python file called `test_pipes.py`. This file relies on `shutil.which()` to identify the path to the `Rscript` binary on my machine. You may need to modify your PATH environment variable if `shutil.which()` does not detect your R installation. 

I also had to move my `test.R` script from the root of my project into my module directory in order for `file_relative_path(__file__, "test.R")` to work. 

```python
# dagster_and_r/test_pipes.py

import shutil

from dagster import (
    AssetExecutionContext,
    MaterializeResult,
    PipesSubprocessClient,
    asset,
    file_relative_path,
)


@asset
def subprocess_asset(
    context: AssetExecutionContext,
    pipes_subprocess_client: PipesSubprocessClient,
) -> MaterializeResult:
    cmd = [shutil.which("Rscript"), file_relative_path(__file__, "test.R")]
    return pipes_subprocess_client.run(
        command=cmd,
        context=context,
    ).get_materialize_result()
```

Then, I made sure to import my `subprocess_asset` and define the pipes subprocess client resource within the module's `__init__.py`:

```python
# dagster_and_r/__init__.py
from dagster import (
    Definitions,
    PipesSubprocessClient,
    )

from . import assets
from . test_pipes import subprocess_asset

defs = Definitions(
    assets=[subprocess_asset],
    resources={"pipes_subprocess_client": PipesSubprocessClient()}
)
```
From here, I was able to launch the Dagster UI and materialize my new asset!

![image](https://github.com/dagster-io/dagster/assets/24445174/ed290f8d-d9f5-4798-bce5-9341d7d51391)

**Reply by PedramNavid** [reply]
thanks for the update! i’m sure others will find this useful as well

**Reply by philiporlando** [reply]
Np! Currently trying to pass context between dagster and R via `reticulate`, but it's proving to be extremely difficult to debug... I'll report back if I figure out how to do this.

**Reply by philiporlando** [reply]
Ok, passing context between Dagster and R is a little more complicated, but still possible through the [{reticulate}](https://rstudio.github.io/reticulate/) package. This approach requires `>=reticulate-1.34.0`. 

I've updated my R script to use the `open_dagster_pipes` context manager like this:

```r
# dagster_and_r/dagster_pipes_with_context.R

# I forced the python binary used by reticulate by setting the RETICULATE_PYTHON environment variable
# You may have to manually set this within your `.Renviron` and restart your R session if this doesn't work. 
Sys.setenv(RETICULATE_PYTHON = "path/to/python/environment/with/dagster/python.exe")

# Load reticulate
library(reticulate)

# Check to see if dagster_pipes module is available within the python environment
reticulate::py_module_available("dagster_pipes")

# Handle python imports 
# R doesn't support selective imports like Python, so you have to do this
# to avoid typing the full namespace path repeatedly...
dagster_pipes <- reticulate::import("dagster_pipes")
open_dagster_pipes <- dagster_pipes$open_dagster_pipes
PipesContext <- dagster_pipes$PipesContext


# https://rstudio.github.io/reticulate/articles/calling_python.html#with-contexts
with(open_dagster_pipes() %as% pipes, {
  context <- PipesContext$get()
  context$log$info("This is a log from R!")
  data(iris)
  context$log$info(head(iris))
})
```

Here's my updated asset

```python
# dagster_and_r/dagster_pipes_and_r.py

import shutil

from dagster import (
    AssetCheckResult,
    AssetExecutionContext,
    MaterializeResult,
    PipesSubprocessClient,
    asset,
    file_relative_path,
)

@asset
def my_r_asset_with_context(
    context: AssetExecutionContext,
    pipes_subprocess_client: PipesSubprocessClient,
) -> MaterializeResult:
    cmd = [shutil.which("Rscript"), file_relative_path(__file__, "dagster_pipes_with_context.R")]
    return pipes_subprocess_client.run(
        command=cmd,
        context=context,
    ).get_materialize_result()
```

And here's my definitions:

```python
# dagster_and_r/__init__.py

from dagster import (
    Definitions,
    PipesSubprocessClient,
    )
    
from .dagster_pipes_and_r import (
    my_r_asset_with_context,
    )


defs = Definitions(
    assets=[
        my_r_asset_with_context,
        ],
    resources={"pipes_subprocess_client": PipesSubprocessClient()}
)
```

This materializes without error and the logs executed from my external R session are visible within the Dagster UI!

![image](https://github.com/dagster-io/dagster/assets/24445174/6793c296-5616-489c-8bae-a3ede1551bab)


Next steps are to troubleshoot other features like asset checks, and passing objects/data between Dagster and R. It would be amazing if I could return that `iris` dataset and then perform some transformation to it using a Python data frame library like polars or something! Fingers crossed that this is possible!

**Reply by PedramNavid** [reply]
Wow! This is blog-post-worthy work Philip.

**Reply by philiporlando** [reply]
@PedramNavid, thanks! I'll work on at least standing up a git repo to track this work.

**Reply by philiporlando** [reply]
Ok, here's an update re passing data between R and Dagster:

> Dagster Pipes does not currently support streaming data objects directly between external processes and Dagster assets. - Scout

Since Dagster Pipes does not currently support in-memory data passing between the external process and Dagster, I don't think it will be possible to decouple IO from business logic when defining assets that execute external R code. Not even `reticulate::r_to_py()` can save us here! Instead, we'd need to serialize the data to an intermediate storage system that both languages can understand. For example, the `iris` data.frame could be serialized to CSV, JSON, Parquet, Arrow, DuckDB, etc., and then have a downstream Python asset read and deserialize the data from that file. 

```r
write.csv(iris, "iris.csv", row.names = FALSE)
```

```python
import pandas as pd

@asset
def my_python_asset():
    df = pd.read_csv("iris.csv")
    # Perform operations on the dataframe
    return df
```

Despite this limitation, I still feel like Dagster Pipes is the best way to execute R code from Dagster because you can still pass context, environment variables, logs, and metadata back and forth, which gives me way more visibility about the R asset that is being produced-- and all from within the Dagster UI. I'd call that a huge win!

**Reply by alangenfeld** [reply]
>  in-memory data passing

While its not supported in pipes directly, it might be possible to 

1. create a `SharedMemory` on the orchestration side
2. pass the `name` using pipes `extras` / env var / etc.
3. open the `SharedMemory`  by name in the `R` process using the `name` and write data in to it
4. read the data out on the orchestration side when the external process finishes

Some limitations could cause this path to be infeasible, but could be worth exploring 

https://docs.python.org/3/library/multiprocessing.shared_memory.html

**Reply by philiporlando** [reply]
Thanks for the suggestion, @alangenfeld! I'll give this a shot and see what happens.

**Reply by philiporlando** [reply]
I created [this repo](https://github.com/philiporlando/dagster-and-r) to further delve into the integration of Dagster and R. It includes some examples of using asset checks with R. I'll be sure to include any other progress I make with in-memory data passing there too.
